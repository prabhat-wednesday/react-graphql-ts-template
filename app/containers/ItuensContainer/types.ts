export interface Song {
  trackId: number;
  artistName: string;
  artworkUrl100: string;
  collectionName: string;
}

export interface SongData {
  Songs: Song[];
}
