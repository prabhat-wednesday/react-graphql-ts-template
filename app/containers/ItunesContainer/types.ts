import { AnyAction } from '@reduxjs/toolkit';

export interface Song {
  trackId: number;
  artistName: string;
  artworkUrl100: string;
  collectionName: string;
  previewUrl: string;
}

export interface SongData {
  results: Song[];
}

export type SongActionCreator = (payload: string) => AnyAction;

export interface ItuneContainerProps {
  dispatchArtistName: SongActionCreator;
  songData: {
    results: Song[];
  };
  loading: boolean;
  songListError?: string;
}
