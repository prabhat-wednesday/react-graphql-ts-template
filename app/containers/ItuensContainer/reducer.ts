import { createSlice } from '@reduxjs/toolkit';

export interface ApiResponseState {
  loading: boolean;
  dataToShow: [];
  error: string;
  searchTerm: string;
}

export const initialState: ApiResponseState = {
  loading: false,
  dataToShow: [],
  error: '',
  searchTerm: ''
};

const ituneSlice = createSlice({
  name: 'Itune',
  initialState,
  reducers: {
    getSearchTerm: (state, action) => {
      state.searchTerm = action.payload;
    },
    getDataToShow: (state, action) => {
      state.dataToShow = action.payload;
    }
  }
});

export const { getSearchTerm, getDataToShow } = ituneSlice.actions;

export default ituneSlice.reducer;
