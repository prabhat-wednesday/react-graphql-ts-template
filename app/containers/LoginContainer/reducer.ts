import { createSlice } from '@reduxjs/toolkit';

export interface loginProviderState {
  userData: {};
  isLogin: boolean;
}

export const initialState: loginProviderState = {
  userData: {},
  isLogin: false
};

const LoginProviderSlice = createSlice({
  name: 'LoginProviderSlice',
  initialState,
  reducers: {
    requestGetUserData: (state, action) => {
      state.userData = action.payload;
      state.isLogin = true;
    },
    logoutUser: (state) => {
      state.userData = {};
      state.isLogin = false;
    }
  }
});

export const { requestGetUserData, logoutUser } = LoginProviderSlice.actions;

export default LoginProviderSlice.reducer;
