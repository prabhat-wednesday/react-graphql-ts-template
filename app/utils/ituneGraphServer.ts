import ApolloClient, { DocumentNode, InMemoryCache } from 'apollo-boost';

export const client = new ApolloClient({
  uri: 'https://itunes.apple.com/search?term=',
  cache: new InMemoryCache()
});

export interface GqlQueryReponse<Data> {
  data?: Data;
  error?: any;
  ok: boolean;
}

export const getQueryResponse = <Data, Variables>(
  query: DocumentNode,
  variables?: Variables
): Promise<GqlQueryReponse<Data>> => {
  console.log(query, 'in graphql server');
  return client
    .query<Data, Variables>({ query, variables })
    .then((res) => {
      console.log(res, 'in graphql');
      if (res.errors) {
        console.log(res);
        return { error: res.errors, ok: false };
      } else {
        console.log(res);
        return { data: res.data, ok: true };
      }
    })
    .catch((err) => {
      return { error: err, ok: false };
    });
};
