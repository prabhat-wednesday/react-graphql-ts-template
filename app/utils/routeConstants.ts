export type RouteConstant = {
  route: string;
  exact?: boolean;
  isProtected?: boolean;
  props?: object;
};

const routeConstants: Record<string, RouteConstant> = {
  itune: {
    route: '/',
    exact: true
  },
  song: {
    route: '/song/:trackId',
    exact: true
  }
};

export type RouteKeys = keyof typeof routeConstants;

export default routeConstants;
