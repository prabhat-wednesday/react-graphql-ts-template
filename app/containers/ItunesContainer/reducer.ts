import { prepare } from '@app/utils';
import { createSlice } from '@reduxjs/toolkit';
import { song } from './types';

export interface ituneResponseState {
  loading: boolean;
  songData: {
    resultCount: number;
    results: song[];
  };
  songListError: string;
}

export const initialState: ituneResponseState = {
  loading: false,
  songData: { resultCount: 0, results: [] },
  songListError: ''
};

const ituneSlice = createSlice({
  name: 'Itune',
  initialState,
  reducers: {
    requestGetSongList: {
      reducer: (state) => {
        state.loading = true;
      },
      prepare
    },
    successGetSongList: (state, action) => {
      state.songData = action.payload;
      state.songListError = '';
      state.loading = false;
    },
    failureGetSongList: (state, action) => {
      state.songData.resultCount = 0;
      state.songData.results = [];
      state.loading = false;
      state.songListError = action.payload;
    }
  }
});

export const { requestGetSongList, successGetSongList, failureGetSongList } = ituneSlice.actions;

export default ituneSlice.reducer;
