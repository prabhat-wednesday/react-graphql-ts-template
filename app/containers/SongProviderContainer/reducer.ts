import { prepare } from '@app/utils';
import { createSlice } from '@reduxjs/toolkit';
import { Song } from '../ItunesContainer/types';

interface songProviderState {
  songData: {
    results: Song[];
  };
  loading: boolean;
  songListError: string;
}

export const initialState: songProviderState = {
  songData: {
    results: []
  },
  loading: false,
  songListError: ''
};

const SongProviderSlice = createSlice({
  name: 'SongProviderSlice',
  initialState,
  reducers: {
    requestGetSongList: {
      reducer: (state) => {
        state.loading = true;
      },
      prepare
    },
    successGetSongList: (state, action) => {
      state.songData.results = action.payload.results;
      state.songListError = '';
      state.loading = false;
    },
    failureGetSongList: (state, action) => {
      state.songData.results = [];
      state.loading = false;
      state.songListError = action.payload;
    }
  }
});

export const { requestGetSongList, successGetSongList, failureGetSongList } = SongProviderSlice.actions;

export default SongProviderSlice.reducer;
