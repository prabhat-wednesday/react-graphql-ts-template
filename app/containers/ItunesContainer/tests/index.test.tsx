import React from 'react';
import { renderProvider, timeout } from '@app/utils/testUtils';
import { ItunesContainerTest as ItunesContainer, mapDispatchToProps } from '..';
import { requestGetSongList } from '../reducer';
import { ItuneContainerProps } from '../types';
import { fireEvent } from '@testing-library/react';

describe('<ItuneContainer /> test', () => {
  let submitSpy: jest.Mock;
  let defaultProps: ItuneContainerProps;

  beforeEach(() => {
    submitSpy = jest.fn();
    defaultProps = {
      dispatchArtistName: submitSpy,
      loading: true,
      songData: {
        results: [
          {
            trackId: 1,
            artistName: 'Yung Xiety',
            artworkUrl100:
              'https://is1-ssl.mzstatic.com/image/thumb/Music112/v4/3b/cc/31/3bcc315a-e58f-c678-8d05-b7d62346070f/1d724fe3-daaa-4825-91f0-a47e70a438bd.jpg/100x100bb.jpg',
            collectionName: 'Arijit Singh (Mashup) - Single'
          }
        ]
      }
    };
  });

  it('should render and match the snapshot', () => {
    const { baseElement, getByTestId } = renderProvider(<ItunesContainer {...defaultProps} />);
    expect(baseElement).toMatchSnapshot();
    expect(getByTestId('search-label')).toBeInTheDocument();
  });

  it('should validate the dispatchArtistName action', () => {
    const dispatchArtistNameSpy = jest.fn();
    const artistName = 'Arijit Singh';
    const action = {
      dispatchArtistName: requestGetSongList(artistName)
    };
    const props = mapDispatchToProps(dispatchArtistNameSpy);
    props.dispatchArtistName(artistName);
    expect(dispatchArtistNameSpy).toHaveBeenCalledWith(action.dispatchArtistName);
  });

  it('should call the dispatchArtistName on change', async () => {
    const { getByTestId } = renderProvider(
      <ItunesContainer {...defaultProps} dispatchArtistName={submitSpy} loading={false} />
    );
    fireEvent.change(getByTestId('search-bar'), {
      target: { value: 'Arijit Singh' }
    });
    await timeout(500);
    expect(submitSpy).toBeCalledWith('Arijit Singh');
  });
});
