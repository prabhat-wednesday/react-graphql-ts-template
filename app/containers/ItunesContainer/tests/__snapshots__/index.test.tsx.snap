// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ItuneContainer /> test should render and match the snapshot 1`] = `
<body>
  <div>
    <div>
      <div
        class="ItunesContainer__InputContainer-sc-1mn0u29-0 cWHxEw"
      >
        <p
          class="T__StyledText-gcrjip-0 gIxZML"
          data-testid="search-label"
        >
          Search your favourite singer
        </p>
        <input
          aria-label="input-element"
          class="ant-input ItunesContainer__CustomInput-sc-1mn0u29-1 gSSPZi"
          data-testid="search-bar"
          type="text"
          value=""
        />
      </div>
      <div
        class="ant-skeleton ant-skeleton-active"
      >
        <div
          class="ant-skeleton-content"
        >
          <h3
            class="ant-skeleton-title"
            style="width: 38%;"
          />
          <ul
            class="ant-skeleton-paragraph"
          >
            <li />
            <li />
            <li
              style="width: 61%;"
            />
          </ul>
        </div>
      </div>
    </div>
  </div>
</body>
`;
